illumos-utils/src/host/executor.rs

```diff
@@ -0,0 +1,532 @@
+    }
+}
+
+type DynamicHandler = Box<dyn FnMut(Input) -> Output + Send + Sync>;
+
+enum HandledCommand {
+    Static { input: Input, output: Output },
+    Dynamic { handler: DynamicHandler },
+}
+
+/// A handler that may be used for setting inputs/outputs to the executor
+/// when these commands are known ahead-of-time.
+///
+/// See: [Self::register] for integration with a [FakeExecutor].
+pub struct StaticHandler {
+    expected: Vec<HandledCommand>,
+    index: usize,
+}
+
+impl StaticHandler {
+    pub fn new() -> Self {
+        Self { expected: Vec::new(), index: 0 }
+    }
+
+    /// Convenience function to register the handler with a [FakeExecutor].
+    pub fn register(mut self, executor: &FakeExecutor) {
+        executor.set_wait_handler(Box::new(move |child| -> Output {
+            self.execute(child.command())
+        }));
+    }
+
+    /// Expects a static "input" to exactly produce some "output".
+    pub fn expect(&mut self, input: Input, output: Output) {
+        self.expected.push(HandledCommand::Static { input, output });
+    }
+
+    /// A helper for [Self::expect] which quietly succeeds.
+    pub fn expect_ok<S: AsRef<str>>(&mut self, input: S) {
+        self.expect(Input::shell(input), Output::success())
+    }
+
+    /// A helper for [Self::expect] which quietly fails.
+    pub fn expect_fail<S: AsRef<str>>(&mut self, input: S) {
+        self.expect(Input::shell(input), Output::failure())
+    }
+
+    /// Expects a dynamic handler to be invoked to dynamically
+    /// determine the output of this call.
+    pub fn expect_dynamic(&mut self, handler: DynamicHandler) {
+        self.expected.push(HandledCommand::Dynamic { handler });
+    }
+
+    fn execute(&mut self, command: &Command) -> Output {
+        let observed_input = Input::from(command);
+        let expected = &mut self
+            .expected
+            .get_mut(self.index)
+            .unwrap_or_else(|| panic!("Unexpected command: {observed_input}"));
+        self.index += 1;
+
+        match expected {
+            HandledCommand::Static { input, output } => {
+                assert_eq!(&observed_input, input, "Unexpected input command");
+                output.clone()
+            }
+            HandledCommand::Dynamic { ref mut handler } => {
+                handler(observed_input)
+            }
+        }
+    }
+}
+
+impl Drop for StaticHandler {
+    fn drop(&mut self) {
+        let expected = self.expected.len();
+        let actual = self.index;
+        if actual < expected {
+            let next = &self.expected[actual];
+            let tip = match next {
+                HandledCommand::Static { input, .. } => input.to_string(),
+                HandledCommand::Dynamic { .. } => {
+                    "<dynamic handler>".to_string()
+                }
+            };
+            let errmsg = format!("Only saw {actual} calls, expected {expected}\nNext would have been: {tip}");
+            if !std::thread::panicking() {
```

Comment:
> So in a case where a caller creates a CommandSequence, and does not call that assert_all_expected_commands_seen function, would we just silently succeed? Is that the desired behavior?

That'd be my proposal, yeah. I don't love it, but still prefer it to a `drop` check (even one that's just for tests). This is not a hill I'm dying on, so if you prefer the drop check and want to keep it I can certainly live with that. The only other alternative I can think of is to do something with a closure like

```rust
let mut handler = CommandSequence::new();
// add handler.expect_* expectations

handler.with_executor(log, |executor| {
    // ... actual tests ...
});
```

where `with_executor` would do these assertions after calling the closure. No `drop` involved; if the tests fail/panic the assertions associated with the handler don't run (which is what we want in general, I think?). But the rightward drift of having to wrap all the actual tests in a closure is also annoying, so this doesn't seem like a surefire winner either.

Reply:
For now, I do mildly prefer the drop check, but I'd be willing to change it when I get a better grasp on usage?

I'm slightly nervous about the closure-based approach, because moving things into / out of an async context is painful under test, and it's frustrating for tests that might want to use multiple executors.